op-service/sources/prefetching_eth_client.go

```diff
@@ -0,0 +1,238 @@
+package sources
+
+import (
+	"context"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type EthClientInterface interface {
+	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error)
+	ChainID(ctx context.Context) (*big.Int, error)
+	InfoByHash(ctx context.Context, hash common.Hash) (eth.BlockInfo, error)
+	InfoByNumber(ctx context.Context, number uint64) (eth.BlockInfo, error)
+	InfoByLabel(ctx context.Context, label eth.BlockLabel) (eth.BlockInfo, error)
+	InfoAndTxsByHash(ctx context.Context, hash common.Hash) (eth.BlockInfo, types.Transactions, error)
+	InfoAndTxsByNumber(ctx context.Context, number uint64) (eth.BlockInfo, types.Transactions, error)
+	InfoAndTxsByLabel(ctx context.Context, label eth.BlockLabel) (eth.BlockInfo, types.Transactions, error)
+	PayloadByHash(ctx context.Context, hash common.Hash) (*eth.ExecutionPayload, error)
+	PayloadByNumber(ctx context.Context, number uint64) (*eth.ExecutionPayload, error)
+	PayloadByLabel(ctx context.Context, label eth.BlockLabel) (*eth.ExecutionPayload, error)
+	FetchReceipts(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Receipts, error)
+	GetProof(ctx context.Context, address common.Address, storage []common.Hash, blockTag string) (*eth.AccountResult, error)
+	GetStorageAt(ctx context.Context, address common.Address, storageSlot common.Hash, blockTag string) (common.Hash, error)
+	ReadStorageAt(ctx context.Context, address common.Address, storageSlot common.Hash, blockHash common.Hash) (common.Hash, error)
+	Close()
+}
+
+type PrefetchingEthClient struct {
+	inner                 EthClientInterface
+	PrefetchingRange      uint64
+	PrefetchingTimeout    time.Duration
+	runningCtx            context.Context
+	runningCancel         context.CancelFunc
+	highestHeadRequesting uint64
+	highestHeadLock       sync.Mutex
+}
+
+// NewPrefetchingEthClient creates a new [PrefetchingEthClient] with the given underlying [EthClient]
+// and a prefetching range.
+func NewPrefetchingEthClient(inner EthClientInterface, prefetchingRange uint64, timeout time.Duration) (*PrefetchingEthClient, error) {
+	// Create a new context for the prefetching goroutines
+	runningCtx, runningCancel := context.WithCancel(context.Background())
+	return &PrefetchingEthClient{
+		inner:                 inner,
+		PrefetchingRange:      prefetchingRange,
+		PrefetchingTimeout:    timeout,
+		runningCtx:            runningCtx,
+		runningCancel:         runningCancel,
+		highestHeadRequesting: 0,
+	}, nil
+}
+
+func (p *PrefetchingEthClient) FetchWindow(start, end uint64) {
+	// Acquire lock before reading/updating highestHeadRequesting
+	p.highestHeadLock.Lock()
+	if start <= p.highestHeadRequesting {
+		start = p.highestHeadRequesting + 1
+	}
+	if p.highestHeadRequesting <= end {
```

Comment:
nit: works either way, but update to `highestHeadRequesting` only needed in strict unequal case